Clean/Gadgets/Addition8Full.lean

```diff
@@ -0,0 +1,244 @@
+import Clean.GenericConstraint
+import Clean.Expression
+import Clean.Gadgets.Boolean
+import Clean.Gadgets.ByteLookup
+import Clean.Utils.Field
+import Mathlib.Algebra.Field.Basic
+import Mathlib.Data.ZMod.Basic
+
+/-
+  8-bit full adder gadget: the output `out` must be the sum of the
+  input bytes `x` and `y` plus the `carry_in` boolean modulo 256,
+  and the output carry `carry_out` must be the quotient
+  of the sum of `x`, `y` and `carry_in` divided by 256.
+-/
+namespace Addition8Full
+open Expression
+variable {p : ℕ} [p_is_prime: Fact p.Prime] [p_large_enough: Fact (p > 512)]
+
+def circuit (N : ℕ+) (M : ℕ) (x y out carry_in carry_out : Expression N M (F p)) : ConstraintGadget p N M :=
+  ⟨
+    [
+      carry_in + x + y - out - carry_out * (const 256)
+    ],
+    [
+      ByteLookup.lookup N M x,
+      ByteLookup.lookup N M y,
+      ByteLookup.lookup N M out,
+    ],
+    [
+      Boolean.circuit N M carry_in,
```

Comment:
I guess when passing carries from one to another instance of this gadget, they will be constrained twice -- once as `carry_in`, once as `carry_out`